### PR TITLE
feat: daily bridge transfer limits

### PR DESCRIPTION
**Pull request by maharifu**
# Related Github tickets

- Closes https://github.com/VolumeFi/paloma/issues/1711

# Background

Add a limit to the amount of tokens transferred out from paloma within a specified window, per token. This is defined by a new governance vote.

- Add governance vote. Can define token, window size, token limit and list of exempt addresses.
- Keep a counter of the number of tokens transferred and the block height at the start of the limiting window.
- Before accepting a transfer, update the counters and check the limits.
- Transfers are refused if over the transfer limit.

# Testing completed

- [x] test coverage exists or has been added/updated
- [x] tested in a private testnet

# Breaking changes

- [x] I have checked my code for breaking changes
- [x] If there are breaking changes, there is a supporting migration.

### CODE REVIEW

**Comment by maharifu** [comment]
@taariq @byte-bandit I assumed we want to update/check the counters when adding the transfer to the queue, and will not decrease the counters even if the transfer fails or is cancelled afterwards.

**Comment by taariq** [comment]
@maharifu correct. No need to decrease the counters on cancellation. All good.

**Comment by taariq** [comment]
@byte-bandit your review? Let's merge this in!

File: proto/palomachain/paloma/gravity/query.proto [hunk 1]
```diff
@@ -48,6 +49,12 @@ service Query {
   rpc GetBridgeTax(google.protobuf.Empty) returns (QueryBridgeTaxResponse) {
     option (google.api.http).get = "/palomachain/paloma/gravity/bridge_tax";
   }
+
+  rpc GetAllBridgeTransferLimits(google.protobuf.Empty)
```
Comment (byte-bandit) [inline proto/palomachain/paloma/gravity/query.proto:53]:
```suggestion
  rpc GetBridgeTransferLimits(google.protobuf.Empty)
```

File: x/gravity/client/cli/query.go [hunk 2]
```diff
@@ -366,3 +367,33 @@ func GetCmdQueryBridgeTax() *cobra.Command {
 
 	return cmd
 }
+
+// GetCmdQueryAllBridgeTransferLimits fetches transfer limits for all tokens
+func GetCmdQueryAllBridgeTransferLimits() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "all-bridge-transfer-limits",
```
Comment (byte-bandit) [inline x/gravity/client/cli/query.go:374]:
```suggestion
		Use:   "bridge-transfer-limits",
```

File: x/gravity/keeper/grpc_query_get_all_bridge_transfer_limits.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	"context"
+	"errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	keeperutil "github.com/palomachain/paloma/util/keeper"
+	"github.com/palomachain/paloma/x/gravity/types"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+func (k Keeper) GetAllBridgeTransferLimits(
```
Comment (byte-bandit) [inline x/gravity/keeper/grpc_query_get_all_bridge_transfer_limits.go:13]:
```suggestion
func (k Keeper) GetBridgeTransferLimits(
```

Comment (byte-bandit) [inline x/gravity/keeper/grpc_query_get_all_bridge_transfer_limits.go:13]: FYI, I've been thinking a bit about API design on the keepers lately. The most convincing argument I've come across is to 

- start functions related to gRPC queries with `Query...`
- start functions on message server implementation with the verb, i.e. `Relay...`, `Start...`, etc...
- start simple getter functions on the keeper implementation with `Get`
- name more advanced getters by their index, i.e. `Get...ByID`
- start iteration callbacks with `Iterate...`

It's not perfect. Personally, I'd prefer calling getters by their entity, and setters by their function, i.e. `BridgeLimit`, `BridgeLimits`, `BridgeLimitByID`, `UpdateBridgeLimit`, etc...

But the thoughts outlined above seem to be more consistent with the cosmos SDK coding style. Just some food for thought here.

**Review by byte-bandit — DISMISSED** [review]
Good stuff. Some small input, but non-blocking. Up to you.

File: proto/palomachain/paloma/gravity/bridge_transfer_limit_proposal.proto [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+message SetBridgeTransferLimitProposal {
+  string title = 1;
+  string description = 2;
+  string token = 3;
+  string limit = 4 [
+    (gogoproto.customtype) = "cosmossdk.io/math.Int",
+    (gogoproto.nullable) = false
+  ];
+  string limit_period = 5;
```
[the remark this answers is not in the document]
Comment (maharifu) [reply proto/palomachain/paloma/gravity/bridge_transfer_limit_proposal.proto:16]: Nice catch. I was converting it in the keeper, but it makes sense to convert it earlier in the command and return an error then. :+1:

**Review by byte-bandit — APPROVED** [review]

**Pull request merged**